src/dbnode/storage/index/block.go

```diff
@@ -128,32 +215,254 @@ func (b *block) EndTime() time.Time {
 	return b.endTime
 }
 
-func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
+func (b *block) maybeBackgroundCompactWithLock() {
+	if b.compactingBackground || b.state != blockStateOpen {
+		return
+	}
+
+	// Create a logical plan
+	segs := make([]compaction.Segment, 0, len(b.backgroundSegments))
+	for _, seg := range b.backgroundSegments {
+		segs = append(segs, compaction.Segment{
+			Age:     seg.Age(),
+			Size:    seg.Segment().Size(),
+			Type:    segments.FSTType,
+			Segment: seg.Segment(),
+		})
+	}
+
+	plan, err := compaction.NewPlan(segs, b.opts.BackgroundCompactionPlannerOptions())
+	if err != nil {
+		instrument.EmitAndLogInvariantViolation(b.iopts, func(l xlog.Logger) {
+			l.Errorf("index background compaction plan error: %v", err)
+		})
+		return
+	}
+
+	if len(plan.Tasks) == 0 {
+		return
+	}
+
+	// Kick off compaction
+	b.compactingBackground = true
+	go func() {
```

Comment:
Not a huge fan of this pattern where you kick off a goroutine that at the end maybe kicks itself off again after checking some state...

Could the block just have a single-threaded background routine that checks for background segments and compacts them? seems simpler and you already have a `Close()` method for cleanup.

Reply:
You have more context than I do, but after reading through the whole file the flow of operations is kind of hard to track. As far as I can tell, at a given time you can have both forground compactions going on, as well as background compactions.

Not saying we should necessarily change it now, but I wonder if it would be easier to understand this code if it was modeled as a pipeline where the flow is:

Write batch --> Foreground compactor (single threaded) --> Background compactor (single threaded)

and then that would be implemented with two goroutines and some channels.

Reply:
Yeah so the thing is the background compaction loop needs to keep going indefinitely until it really can't anymore.  That's why you can't trigger it necessarily from the foreground goroutine and have it kick of sequentially each time.  

Some times it will be in the middle of a background compaction, and you need to make sure it kicks itself back off again at the end of it's current run.  That's why you have to have when it finishes, it needs to check if it needs to run again or not.

Reply:
Yeah I guess it just seems like you could have some backgroung goroutines and some message passing or something like that, and then the background routine can receive a close signal and do any remaining work before it shuts down if it needs to keep going after a call to Close(), but I'll leave it to you. I'm sure it could be one of those things where you try an alternative that seems cleaner and it actually turns out way worse

Reply:
Yeah I played with a few different ways, and this way was actually cleanest.  Especially since both foreground and compaction could be modeled similarly.